### PR TITLE
[TH] Initial Translation for Concurrency

### PR DESCRIPTION
**Pull request by phoomparin**
I've translated the concurrency lesson in Thai. Thanks!

What do you think, P' @iboss-ptk?

### CODE REVIEW

File: th/lessons/advanced/concurrency.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
+---
+version: 1.0.0
+title: การทำงานพร้อมกัน (Concurrency)
+---
+
+หนึ่งในจุดขายหลักของ Elixir คือการรองรับการทำงานพร้อมๆ กัน ซึ่งเป็นผลพลอยได้จากการนำ Erlang VM (BEAM) มาใช้งาน ส่งผลให้การจัดการการทำงานพร้อมกันใน Elixir นั้นง่ายกว่าที่คิดมากๆ
+
+โดยรูปแบบการทำงานพร้อมๆ กันนั้น มีความจำเป็นต้องพึ่งพา Actors​ ซึ่งเป็นโปรเซสที่สามารถสื่อสารกับโปรเซสอื่นๆ ได้โดยผ่านการส่งข้อความหากัน
```
Comment (zkan) [inline th/lessons/advanced/concurrency.md]: Actors -> actor

Let use the singular term for the entire translation according to @iboss-ptk' s reason here: https://github.com/elixirschool/elixirschool/pull/1194#discussion_r143441228, which I also agree with him. :)

Sounds good?

File: th/lessons/advanced/concurrency.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+ในบทเรียนนี้ เราจะมาลองเรียนรู้ถึงโมดูลที่เอาไว้ใช้จัดการเรื่องการทำงานพร้อมกัน ที่มาพร้อมกับ Elixir ซึ่งในบทต่อๆ ไป เราจะพูดถึงคุณลักษณะต่างๆ ของ OTP ที่นำเรื่องการทำงานพร้อมการมาใช้ด้วย
+
+{% include toc.html %}
+
+## โปรเซส (Processes)
+
+โปรเซสต่างๆ ใน Erlang VM นั้นเบามากๆ และจะถูกนำไปประมวลผลในทุกๆ CPU ในเครื่อง ซึ่งตัวโปรเซสนั้นอาจจะดูเหมือน Thread ที่มีอยู่ตามปกติ แต่จริงๆ แล้วมีความซับซ้อนน้อยกว่ากันมาก และไม่ใช่เรื่องแปลกเลยที่จะมีหลายพันโปรเซสทำงานพร้อมกันอยู่ในโปรแกรมเดียว
```
Comment (zkan) [inline th/lessons/advanced/concurrency.md]: Thread -> thread

File: th/lessons/advanced/concurrency.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    IO.puts(a + b)
+  end
+end
+
+iex> Example.add(2, 3)
+5
+:ok
+```
+
+ถ้าเราต้องการเรียกใช้ฟังก์ชั่นนี้แบบ Asynchronous เราจะต้องใช้คำสั่ง `spawn/3` แทน:
```
Comment (zkan) [inline th/lessons/advanced/concurrency.md]: Asynchronous -> asynchronous

File: th/lessons/advanced/concurrency.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+end
+
+iex> Example.run
+Exit reason: kaboom
+:ok
+```
+
+## Agents
+
+Agents คือ abstraction ของโปรเซสที่ทำงานอยู่ในเบื้องหลัง ซึ่งมีการเก็บสถานะ (State) อยู่ในตัว โดยเราสามารถเข้าถึงมันจากโปรเซสอื่นๆ ภายในโปรแกรมของเรา และสถานะของ Agent ถูกกำหนดให้เป็นค่าที่ถูกส่งกลับมาจากฟังก์ชั่นของเรา:
```
Comment (zkan) [inline th/lessons/advanced/concurrency.md]: * Agents -> agent
* State -> state
* Agent -> agent

File: th/lessons/advanced/concurrency.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+{:ok, #PID<0.65.0>}
+
+iex> Agent.update(agent, fn (state) -> state ++ [4, 5] end)
+:ok
+
+iex> Agent.get(agent, &(&1))
+[1, 2, 3, 4, 5]
+```
+
+เมื่อเราตั้งชื่อให้ Agent แล้ว เราสามารถเรียกมันด้วยชื่อได้เลย แทนที่จะเรียกด้วย PID:
```
Comment (zkan) [inline th/lessons/advanced/concurrency.md]: Agent -> agent

File: th/lessons/advanced/concurrency.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+iex> Agent.start_link(fn -> [1, 2, 3] end, name: Numbers)
+{:ok, #PID<0.74.0>}
+
+iex> Agent.get(Numbers, &(&1))
+[1, 2, 3]
+```
+
+## หน้าที่การทำงาน (Task)
+
+Task เป็นวิธีที่จะช่วยให้เราสามารถเรียกใช้ฟังก์ชั่นได้ในเบื้องหลัง และรับค่าที่ถูกส่งกลับออกมาได้ในภายหลัง ซึ่งมีความจำเป็นอย่างมากเมื่อเราต้องการที่จะจัดการกับการทำงานที่ใช้การประมวลผลเยอะๆ โดยที่ไม่ต้องไปหยุดการทำงานของโปรแกรมหลัก
```
Comment (zkan) [inline th/lessons/advanced/concurrency.md]: Task -> task

File: th/lessons/advanced/concurrency.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+ในบทเรียนนี้ เราจะมาลองเรียนรู้ถึงโมดูลที่เอาไว้ใช้จัดการเรื่องการทำงานพร้อมกัน ที่มาพร้อมกับ Elixir ซึ่งในบทต่อๆ ไป เราจะพูดถึงคุณลักษณะต่างๆ ของ OTP ที่นำเรื่องการทำงานพร้อมการมาใช้ด้วย
+
+{% include toc.html %}
+
+## โปรเซส (Processes)
+
+โปรเซสต่างๆ ใน Erlang VM นั้นเบามากๆ และจะถูกนำไปประมวลผลในทุกๆ CPU ในเครื่อง ซึ่งตัวโปรเซสนั้นอาจจะดูเหมือน Thread ที่มีอยู่ตามปกติ แต่จริงๆ แล้วมีความซับซ้อนน้อยกว่ากันมาก และไม่ใช่เรื่องแปลกเลยที่จะมีหลายพันโปรเซสทำงานพร้อมกันอยู่ในโปรแกรมเดียว
+
+วิธีที่ง่ายที่สุดที่จะสร้างโปรเซสใหม่ คือการใช้คำสั่ง `spawn` ซึ่งสามารถใส่ฟังก์ชั่นแบบไม่มีชื่อหรือมีชื่อเข้าไปก็ได้ และเมื่อเราสร้างโปรเซสใหม่ขึ้นมา เราจะได้รับตัวบ่งบอกตัวตนของโปรเซส (Process Identifier) หรือที่เรียกกันว่า PID กลับมา เพื่อที่จะได้เรียกมันได้ถูกภายในโปรแกรมของเรา
```
Comment (zkan) [inline th/lessons/advanced/concurrency.md]: รับตัวบ่งบอกตัวตนของโปรเซส (Process Identifier) -> รับ process identifier

I think we can use English word here.

File: th/lessons/advanced/concurrency.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+end
+
+iex> spawn(Example, :explode, [])
+#PID<0.66.0>
+
+iex> spawn_link(Example, :explode, [])
+** (EXIT from #PID<0.57.0>) evaluator process exited with reason: :kaboom
+```
+
+ในบางครั้ง เราก็ไม่ต้องการให้โปรเซสที่เชื่อมโยงกันไปทำให้โปรเซสที่ทำงานอยู่พังลง เราจึงต้องดักฟังสัญญาณหยุดการทำงาน ซึ่งเราจะได้รับข้อความออกมาเป็น Tuple แบบนี้: `{:EXIT, from_pid, reason}`
```
Comment (zkan) [inline th/lessons/advanced/concurrency.md]: Tuple -> tuple

File: th/lessons/advanced/concurrency.md (hunk 1 above)
Comment (phoomparin) [reply th/lessons/advanced/concurrency.md]: Sounds great! Thinking about it, singular term feels a lot more natural.

**Comment by phoomparin** [comment]
Thanks for reviewing, P' kan!

**Pull request merged**